### PR TITLE
Add ML-DSA-65 Key generation.

### PR DESCRIPTION
**Pull request by slontis**
This is only here to demonstrate that ML_DSA is progressing (It still needs some clean ups.)
A feature branch will be required.

The key generation algorithm requires a significant portion of the many algorithms present in FIPS 204.

This work is derived from the BoringSSL code located at https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/mldsa/mldsa.cc

Instead of c++ templates it uses an ML_DSA_PARAMS object to store constants such as k & l. This object also stores EVP_MD_CTX objects for shake128 and shake256, so that the EVP_MD objects are pre fetched. This ctx is then passed to any functions that rely on this information.

The ML_DSA_KEY object stores the encoded public and/or private key whenever a key is loaded or generated.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by baentsch — COMMENTED** [review]
Thanks very much for this heads-up (and all the work, of course), @slontis ! This LGTM (and the usefulness/reusability of your SLH-DSA work cannot be denied, either). Please let me know if there's something you'd like me to do to help this effort move along. @t8m: Any chance you could now create a feature branch for this in case more people than just one would cooperate on this?

File: crypto/build.info [hunk 1]
```diff
@@ -6,7 +6,7 @@ SUBDIRS=objects buffer bio stack lhash hashtable rand evp asn1 pem x509 conf \
         siphash sm3 des aes rc2 rc4 rc5 idea aria bf cast camellia \
         seed sm4 chacha modes bn ec rsa dsa dh sm2 dso engine \
         err comp http ocsp cms ts srp cmac ct async ess crmf cmp encode_decode \
-        ffc hpke thread
+        ffc hpke thread ml_dsa
```
Comment (t8m) [inline crypto/build.info:9]: Hm... we use `mlkem` for ML-KEM but I assume that is the outlier here not `ml_dsa`

**Comment by t8m** [comment]
feature/ml-dsa branch created. You can retarget this PR.

**Comment by slontis** [comment]
> Thanks very much for this heads-up (and all the work, of course), @slontis ! This LGTM (and the usefulness/reusability of your SLH-DSA work cannot be denied, either). Please let me know if there's something you'd like me to do to help this effort move along. @t8m: Any chance you could now create a feature branch for this in case more people than just one would cooperate on this?

I should be ok with the base code. Integrating into TLS 1.3 is something that I have no idea how to do. Matt found this paper here.. https://datatracker.ietf.org/doc/html/draft-tls-westerbaan-mldsa-00
So looking into that would be useful.

**Comment by baentsch** [comment]
> Integrating into TLS 1.3 is something that I have no idea how to do. 

That's straightforward and just a few lines of code registering the algorithm(s) if all the rest is available at the same level as SLH-DSA. fwiw, the very same spec level is just landing in oqsprovider: https://github.com/open-quantum-safe/oqs-provider/pull/568/files and the core `openssl` logic for that is (as)sure(d) to work OK by way of the IETF hackathon interop test scripts.

**Comment by mattcaswell** [comment]
>  Matt found this paper here.. https://datatracker.ietf.org/doc/html/draft-tls-westerbaan-mldsa-00

Note that, in spite of the codepoints listed in that draft, IANA have as yet not updated the signaturescheme registry to list those codepoints.

There was quite some debate on the TLS WG list about this draft and the fact that this is not a hybrid scheme.

Meanwhile there is also this (which is hybrid - but also uses codepoints not in the IANA registry yet).

https://datatracker.ietf.org/doc/draft-reddy-tls-composite-mldsa/

**Comment by baentsch** [comment]
> Meanwhile there is also this (which is hybrid - but also uses codepoints not in the IANA registry yet).
> 
> https://datatracker.ietf.org/doc/draft-reddy-tls-composite-mldsa/

This to me is _very far_ away from things we should worry about right now/in this issue. Created https://github.com/openssl/openssl/issues/26121 to track.

**Comment by slontis** [comment]
NIST seems to be a long way away from accepting hybrid schemes.

**Comment by paulidale** [comment]
> NIST seems to be a long way away from accepting hybrid schemes.

NIST's latest is that hybrid schemes are allowed so long as the PQC part is approved.  The other half is essentially ignored.

**Comment by slontis** [comment]
Closed in favor of https://github.com/openssl/openssl/pull/26127 which uses a feature branch

**Comment by mattcaswell** [comment]
> Closed in favor of https://github.com/openssl/openssl/pull/26127 which uses a feature branch

Note by the way that you don't need to close a PR in order to retarget it at a different branch (click on "Edit" next to the title, and then underneath the title you get the option to change the target branch.

**Pull request closed without merging**